README.md
# color-print
A simple function that allows you to print text with colors and formatting to terminal.

## Information
This package was originally created to allow me to to have more control over the colors shown in my cli's output. Colors that aren't supported by luvit's pretty-print theme.

## Installation
Using [lit](https://luvit.io/lit.html):
```
lit install TohruMKDM/color-print
```
Using [git](https://git-scm.com/):
```
git clone https://github.com/TohruMKDM/color-print.git
```

## Syntax
`[r,g,b;r,g,b](string)`
The first set of rgb values controls the foreground and the second set controls the background.
Values default to 0

## Documentation

#### `color_print.print(...)`
Takes a variable amount of arguments and writes them to stdout just like the standard print. The only difference is that it'll check each argument for the [syntax](https://github.com/TohruMKDM/color-print#syntax) and automatically convert them.</br>
Example: `color_print.print('Hello [red](Only this will be red) World!')`

#### `color_print.dump(value, escape, recursive)`
Primary use of this function is to dump table values to a human-readable string. If value is not a table then it behaves the same as tostring with the only difference being is that it will wrap string values in quotes.
Example: `color_print.dump('Hello World') -> 'Hello World'`<br/>
If `escape` is a truthy then it'll display the literal escape sequence<br/>
Example:
```lua
color_print.dump([[
Multi
Line
String
]], true)
-- '\nMulti\nLine\nString\n'
```
The `recursive` param is for as you've probably guessed, recursively dumping tables.

### `color_print.colorize(color, value)`
Returns value wrapped in the ANSI sequence that will give it color when written to stdout
Useful if you want to get a colorized string to manipulate rather than instantly outputting it. </br>
`color` may be a registered color or a string that allows `r, g, b` </br>
Example: `color_print.colorize('red', 'Red String') -> \027[38;2;255;0;0mRed String\027[0m`

### `color_print.colorize:setColor(name, value)`
Registers a color to be used in the colorizer</br>
Example: `color_print.colorize:setColor('yellow', '254, 221, 0')`</br>
Now `color_print.print('[yellow](Yellow String)')` prints yellow text to terminal

## Example
```lua
local color_print = require('color-print')

-- Print a red 'Hello World!' to stdout.
color_print.print('[255,0,0](Hello World!)')

-- Default values, this does the same thing as the previous call.
color_print.print('[255](Hello World!)')

-- If you want to only set the background color then pass '-' to the foreground.
-- Prints 'Hello World!' to stdout with a red background.
color_print.print('[-;255,0,0](Hello World!)')

-- Prints a red colored 'Hello World' with a green background to stdout.
color_print.print('[255,0,0;0,255,0](Hello World!)')

-- 2.0
-- Colors red, green, and blue are pre-set.
color_print.print('[red](This text will be red)')
color_print.print('[green](This text will be green)')
color_print.print('[blue](This text will be blue)')
-- New formatter
print(color_print.formatter.bold('This text will be bold'))
-- Registers the color yellow as 254, 221, 0
color_print.colorizer:setColor('yellow', '254,221,0')
color_print.print('[yellow](This text will be yellow)')
```

CHANGELOG.md
# Changelog

## 2.0.3
- Fix a small oversight in `formatter.lua`

## 2.0.2
- Made formatter a callable table. `formatter.bold('example') == formatter('bold', 'example')`

## 2.0.1
- Fix issue in package.lua (thanks lit)

## 2.0.0
- Rewrote the package
- Added `dump` to aid in printing large tables
- Added `colorize` which allows you to get colored strings without printing them
- Added `formatter` which allows you to format text such as underlining or boldening
- Improved the printing algorithm to allow printing colored text alongside uncolored text

## 1.0.2
- Fixed a small oversight in `color-print.lua`

## 1.0.1
- Initial release

## 1.0.0
- Published wrong package to lit.

libs/colorize.lua
local format, gmatch = string.format, string.gmatch

local stmtFG = '\027[38;2;%d;%d;%sm%s\027[0m'
local stmtBG = '\027[48;2;%d;%d;%dm%s\027[0m'
local stmtFGBG = '\027[38;2;%d;%d;%d;48;2;%d;%d;%dm%s\027[0m'

local colorize = {}
colorize.__index = colorize

function colorize:__call(color, input)
    color = self.colors[color] or color
    local gen1 = gmatch(color, '[^;]+')
    local FG, BG = gen1(), gen1()
    FG = FG ~= '-' and FG
    if not FG and BG then
        gen1 = gmatch(BG, '[^, ]+')
        return format(stmtBG, gen1() or 0, gen1() or 0, gen1() or 0, input)
    end
    if FG and not BG then
        gen1 = gmatch(FG, '[^, ]+')
        return format(stmtFG, gen1() or 0, gen1() or 0, gen1() or 0, input)
    end
    gen1 = gmatch(FG, '[^, ]+')
    local gen2 = gmatch(BG, '[^, ]+')
    return format(stmtFGBG, gen1() or 0, gen1() or 0, gen1() or 0, gen2() or 0, gen2() or 0, gen2() or 0, input)
end

function colorize:setColor(name, value)
    self.colors[name] = value
end

local colors = {
    red = '255, 0, 0',
    green = '0, 255, 0',
    blue = '0, 0, 255'
}

return setmetatable({colors = colors}, colorize)

libs/dump/init.lua
-- Slight rewrite of the dump function from luvit/pretty-print

local controls = require('./controls')

local match, find = string.match, string.find
local gsub, rep = string.gsub, string.rep
local byte = string.byte
local concat, insert = table.concat, table.insert
local type = type

local dump
local seen, output, stack = {}, {}, {}
local offset = 0

local function escape(input)
    return controls[byte(input, 1)]
end

local function recalculateOffset(index)
    for i = index + 1, #output do
        local value = output[i]
        local result = match(value, '\n([^\n]*)$')
        offset = result and #result or offset + #value
    end
end

local function write(input)
    local length = #input
    local i = 1
    local entry = stack[i]
    while offset + length > 80 and entry do
        if not entry.opened then
            local index = entry.index
            entry.opened = true
            insert(output, index + 1, '\n'..rep(' ', i))
            recalculateOffset(index)
            for x = i + 1, #stack do
                stack[x].index = stack[x].index + 1
            end
        end
        i = i + 1
        entry = stack[i]
    end
    output[#output + 1] = input
    offset = offset + length
end

local process
function process(input, esc, recursive)
    local t = type(input)
    if t == 'string' then
        if find(input, "'") and not find(input, '"') then
            write('"')
            write(esc and gsub(input, '[%c\\\128-\255]', escape) or input)
            write('"')
        else
            write("'")
            write(esc and gsub(input, '[%c\\\128-\255]', escape) or input)
            write("'")
        end
        return
    end
    if t == 'table' and not seen[input] then
        if not recursive then
            seen[input] = true
        end
        write('{ ')
        local i, nextIndex = 1, 1
        for k, v in pairs(input) do
            stack[#stack + 1] = {index = #output, opened = false}
            if k == nextIndex then
                nextIndex = k + 1
                process(v, esc, recursive)
            else
                if type(k) == 'string' and find(k, '^[%a_][%a%d_]*$') then
                    write(k)
                    write(' = ')
                else
                    write('[')
                    process(k, esc, recursive)
                    write(']')
                    write(' = ')
                end
                if type(v) == 'table' then
                    process(v, esc, recursive)
                else
                    stack[#stack + 1] = {index = #output, opened = false}
                    process(v, esc, recursive)
                    stack[#stack] = nil
                end
            end
            write(', ')
            i = i + 1
            stack[#stack] = nil
        end
        output[#output] = ' '
        write('}')
    else
        write(tostring(input))
    end
end

function dump(input, esc, recursive)
    process(input, esc, recursive)
    local result = concat(output)
    seen, output, stack = {}, {}, {}
    offset = 0
    return result
end

return dump

libs/formatter.lua
local format = string.format
local concat = table.concat
local select, tostring = select, tostring

local bold = '\027[1m%s\027[0m'
local faint = '\027[2m%s\027[0m'
local underline = '\027[4m%s\027[0m'
local blink = '\027[5m%s\027[0m'

local formatter = setmetatable({}, {
    __call = function(self, fmt, ...)
        return self[fmt](...)
    end
})

function formatter.bold(...)
    local n = select('#', ...)
    local args = {...}
    for i = 1, n do
        args[i] = tostring(args[i])
    end
    return format(bold, concat(args, '\t'))
end

function formatter.faint(...)
    local n = select('#', ...)
    local args = {...}
    for i = 1, n do
        args[i] = tostring(args[i])
    end
    return format(faint, concat(args, '\t'))
end

function formatter.underline(...)
    local n = select('#', ...)
    local args = {...}
    for i = 1, n do
        args[i] = tostring(args[i])
    end
    return format(underline, concat(args, '\t'))
end

function formatter.blink(...)
    local n = select('#', ...)
    local args = {...}
    for i = 1, n do
        args[i] = tostring(args[i])
    end
    return format(blink, concat(args, '\t'))
end

return formatter

libs/dump/controls.lua
return { [0] = '\\000', '\\001', '\\002', '\\003', '\\004', '\\005', '\\006', '\\a', '\\b', '\\t', '\\n', '\\v', '\\f', 
  '\\r', '\\014', '\\015', '\\016', '\\017', '\\018', '\\019', '\\020', '\\021', '\\022', '\\023', '\\024', '\\025', '\\026', 
  '\\027', '\\028', '\\029', '\\030', '\\031', [34] = '\\"', [39] = "\\'", [92] = '\\\\',
  [128] = '\\128', [129] = '\\129', [130] = '\\130', [131] = '\\131', [132] = '\\132', 
  [133] = '\\133', [134] = '\\134', [135] = '\\135', [136] = '\\136', [137] = '\\137', 
  [138] = '\\138', [139] = '\\139', [140] = '\\140', [141] = '\\141', [142] = '\\142', 
  [143] = '\\143', [144] = '\\144', [145] = '\\145', [146] = '\\146', [147] = '\\147', 
  [148] = '\\148', [149] = '\\149', [150] = '\\150', [151] = '\\151', [152] = '\\152', 
  [153] = '\\153', [154] = '\\154', [155] = '\\155', [156] = '\\156', [157] = '\\157', 
  [158] = '\\158', [159] = '\\159', [160] = '\\160', [161] = '\\161', [162] = '\\162', 
  [163] = '\\163', [164] = '\\164', [165] = '\\165', [166] = '\\166', [167] = '\\167', 
  [168] = '\\168', [169] = '\\169', [170] = '\\170', [171] = '\\171', [172] = '\\172', 
  [173] = '\\173', [174] = '\\174', [175] = '\\175', [176] = '\\176', [177] = '\\177', 
  [178] = '\\178', [179] = '\\179', [180] = '\\180', [181] = '\\181', [182] = '\\182', 
  [183] = '\\183', [184] = '\\184', [185] = '\\185', [186] = '\\186', [187] = '\\187', 
  [188] = '\\188', [189] = '\\189', [190] = '\\190', [191] = '\\191', [192] = '\\192', 
  [193] = '\\193', [194] = '\\194', [195] = '\\195', [196] = '\\196', [197] = '\\197', 
  [198] = '\\198', [199] = '\\199', [200] = '\\200', [201] = '\\201', [202] = '\\202', 
  [203] = '\\203', [204] = '\\204', [205] = '\\205', [206] = '\\206', [207] = '\\207', 
  [208] = '\\208', [209] = '\\209', [210] = '\\210', [211] = '\\211', [212] = '\\212', 
  [213] = '\\213', [214] = '\\214', [215] = '\\215', [216] = '\\216', [217] = '\\217', 
  [218] = '\\218', [219] = '\\219', [220] = '\\220', [221] = '\\221', [222] = '\\222', 
  [223] = '\\223', [224] = '\\224', [225] = '\\225', [226] = '\\226', [227] = '\\227', 
  [228] = '\\228', [229] = '\\229', [230] = '\\230', [231] = '\\231', [232] = '\\232', 
  [233] = '\\233', [234] = '\\234', [235] = '\\235', [236] = '\\236', [237] = '\\237', 
  [238] = '\\238', [239] = '\\239', [240] = '\\240', [241] = '\\241', [242] = '\\242', 
  [243] = '\\243', [244] = '\\244', [245] = '\\245', [246] = '\\246', [247] = '\\247', 
  [248] = '\\248', [249] = '\\249', [250] = '\\250', [251] = '\\251', [252] = '\\252', 
  [253] = '\\253', [254] = '\\254', [255] = '\\255' }

color-print.lua
--[[lit-meta
	name = "TohruMKDM/color-print"
    version = "1.0.1"
    description = "Write text in terminal with colors"
    tags = { "colors", "print", "color" }
    license = "MIT"
    author = { name = "Tohru~ (トール)", email = "admin@ikaros.pw" }
    homepage = "https://github.com/TohruMKDM/color-print"
]]

local sub, format = string.sub, string.format
local match, gmatch = string.match, string.gmatch
local concat = table.concat

local stmtFG = '\027[38;2;%s;%s;%sm%s\027[0m'
local stmtBG = '\027[48;2;%s;%s;%sm%s\027[0m'
local stmtFGBG = '\027[38;2;%s;%s;%s;48;2;%s;%s;%sm%s\027[0m'

return function(...)
    local args = {...}
    for i = 1, #args do
        local arg = tostring(args[i])
        local colors, str = match(arg, '(%b[])(%b())')
        if not colors then
            args[i] = arg
        else
            local gen1, gen2
            colors = sub(colors, 2, -2)
            gen1 = gmatch(colors, '[^;]+')
            local fg, bg = gen1(), gen1()
            fg = fg ~= '-' and fg
            if not fg and bg then
                gen1 = gmatch(bg, '[^, ]+')
                args[i] = format(stmtBG, gen1() or 0, gen1() or 0, gen1() or 0, str)
            elseif fg and not bg then
                gen1 = gmatch(fg, '[^, ]+')
                args[i] = format(stmtFG, gen1() or 0, gen1() or 0, gen1() or 0, str)
            else
                gen1, gen2 = gmatch(fg, '[^, ]+'), gmatch(bg, '[^, ]+')
                args[1] = format(stmtFGBG, gen1() or 0, gen1() or 0, gen1() or 0, gen2() or 0, gen2() or 0, gen2() or 0, str)
            end
        end
    end
    return print(concat(args, '\t'))
end
